### ISSUE TITLE
add syntactic sugar to console.log

### ISSUE THREAD

**Issue by Redsandro** [openhab/openhab-js#50]
TL;DR in **bold**.

The helpful way to initialize Nashorn ECMAScript 5 loggers is like so:

```bash
// A rule
var logger = Java.type('org.slf4j.LoggerFactory')
  .getLogger('org.openhab.rule.' + ctx.ruleUID)
  
// A script
var logger = Java.type('org.slf4j.LoggerFactory')
  .getLogger('org.openhab.script.' + ctx.ruleUID)
```

When a rule calls another rule which calls a script (e.g. using `runNow()`), the logs would show up like so:

```
[INFO ] [org.openhab.rule.e35c19cb69         ]
[INFO ] [org.openhab.rule.b4bd4ca609         ]
[INFO ] [org.openhab.script.1237d80cb9       ]
```

This is very convenient, because when debugging GUI rules and scripts, you can first see if you need to go to the "rules" or "scripts" page, and second you can search for a couple of characters in the UID to open said script without even beginning a thought process about pinpointing the relevant script.

In GraalJS ECMAScript 11, using `console.log()`, the logs become increasingly unhelpful:

```
[INFO ] [org.openhab.automation.script       ]
[INFO ] [org.openhab.automation.script       ]
[INFO ] [org.openhab.automation.script       ]
```

Arguably only the _least relevant part_ of the label remains, and the _two_ most helpful variables are cut.

**Can we add some syntactic sugar to `console.log()` so it will differentiate between GUI rules and scripts and automatically add the relevant rule UID?**

If the Reverse-DNS dot notation with the additional subdomain is crucial while keeping the length within that bracketed space, I suggest shortening it like one of these:

```
[INFO ] [org.openhab.automn.script.1237d80cb9]
[INFO ] [org.openhab.autom.script.1237d80cb9 ]
[INFO ] [org.openhab.auto.script.1237d80cb9  ]
[INFO ] [org.openhab.aut.script.1237d80cb9   ]
[INFO ] [org.openhab.a8n.script.1237d80cb9   ]
```

Keep in mind that the important parts are to the right, and it makes sense to do concessions on the left.

---

PS - to whom it may concern - I noticed this link doesn't work in `README.md` when reading about `console.log()`:

https://github.com/openhab/openhab-js/blob/9c2b0d830d6361cf785ca2b342b8a0d483bb51c3/README.md?plain=1#L101

**Comment by florian-h05** [comment]
> The helpful way to initialize Nashorn ECMAScript 5 loggers is like so: [...]

You can initialize a custom logger in your ECMA 11 script (I think that helps):
```javascript
// A rule
var logger = Java.type('org.slf4j.LoggerFactory')
  .getLogger('org.openhab.rule.' + this.ruleUID);

// A script
var logger = Java.type('org.slf4j.LoggerFactory')
  .getLogger('org.openhab.script.' + this.ruleUID);
```

> Can we add some syntactic sugar to console.log() so it will differentiate between GUI rules and scripts and automatically add the relevant rule UID?

The `console` object is not provided by the openhab-js library, it is provided by the JS Scripting add-on itself, therefore changes have to be made on the add-on.

@digitaldan @jpg0 
Is it be possible to add the uid with `this.ruleUID` to the logger of `console`?

**Comment by digitaldan** [comment]
So i think this is a great idea, and yes it belongs in the jsscripting binding, not here, but thats ok.  The ruleUID is injected in UI configured rules for sure, but i'm not sure about text file based rules (the global injections slightly differ).  In any case, we can at least test to see if its available and append it.  We definitely want to encourage using the console logger over using the raw Java.type as much as possible, also this library supplies a wrapper custom logs already, see https://github.com/openhab/openhab-js#log , but i would also like to encourage using console logging over this as well.

**Comment by jpg0** [comment]
Regarding the UID, I believe that it is autogenerated for text-based rules, based on the rule context (like the name, etc).

It needs to remain the same so that when rules are reloaded they replace the existing rule rather than adding a new one (that is unless we switch to unmanaged rules by default for scripts).

**Comment by jpg0** [comment]
Oh, and regarding logger names: I always used `script.<language>.<filename>` for my code, as I found this the most concise - I agree that they should differ. One question is when to use the ruleUID versus the filename - I find for file rules it's easier with the filename. Possible a format ending with `.<filename>|ui.<ruleUID>` would be best.

Also note that I dropped the `org.openhab.` for script rules. I think that it's debatable whether rule code should go under this namespace or not. 3rd party code that runs inside openHAB does not typically go under this namespace, so I don't really see why script code shouldn't also use it's own namespace. That is why I removed this namespacing from my logger names and switched to a much shorter one. Having something distinct from the standard openHAB namespace makes it much easier to visually distinguish *your* logging from that of the system itself.

**Comment by Redsandro** [comment]
> One question is when to use the ruleUID versus the filename 

New users without prior knowledge see and are encouraged by documentation to use the UI for scripts and rules. They might not even know they can have actual files until they stumble upon a different workflow. For that reason it makes sense to use this ruleUID in logging if the user did not explicitly use a file.

> I find for file rules it's easier with the filename. Possible a format ending with `.<filename>|ui.<ruleUID>` would be best.

That sounds like a sane compromise.

>  I dropped the `org.openhab.` for script rules.

I wanted to do that too (or use use `oh.rs.` for `openhab.redsandro.` to distinguish scripts I borrowed from the community) because when it's everywhere it might as well be nowhere in my opinion. However I think in the ES5 Nashorn implementation I could only use specific namespaces. I don't know if it was a fluke or it is indeed programmatically enforced to use `org.openhab.`

You could have a default of `aut`:

```
[INFO ] [oh.aut.script.ui.1237d80cb9         ]
[INFO ] [oh.aut.rule.ui.e35c19cb69           ]
[INFO ] [oh.aut.script.filename.js           ]
```

And make the log namespace configurable in the script using `something.namespace('rs')`:

```
[INFO ] [oh.rs.script.ui.1237d80cb9          ]
[INFO ] [oh.rs.rule.ui.e35c19cb69            ]
```

**Comment by digitaldan** [comment]
FYI , script based files do inject the script file path as a global variable `javax.script.filename` , so we can definitely make this better.

**Comment by digitaldan** [comment]
So ruleUID is only available in UI based rules.   

I have the following logic to either use the `ruleUID` if found OR parse out the file name from the script path and use that (needs to support windows and unix, hence the regex). 

```javascript
    const scriptId = globalThis["ruleUID"] !== undefined ? ".ui." + globalThis["ruleUID"] : globalThis["javax.script.filename"] !== undefined ? ".file." + globalThis["javax.script.filename"].replace(/^.*[\\\/]/, '') : "";
```

 This prepends `.ui.` for UI based rules and `.file.` for file based ones.  This `scriptId` would then be appended to `org.openhab.automation.script`,  for example:

`[nhab.automation.script.ui.dae9651cb4]` 
or 
`[enhab.automation.script.file.test.js]`

Thoughts ?

**Comment by digitaldan** [comment]
see https://github.com/openhab/openhab-addons/pull/11945 , please feel free to leave comments or suggestions.

**Comment by florian-h05** [comment]
Thank you for your work, LGTM!

The solution with `ui.` + `ruleUID `and `file.` + `filename` is perfect, the logger ids are absolutely clear.

**Comment by digitaldan** [comment]
Tracking this at openhab/openhab-addons#11945
